lib/algae/result.ex

```diff
@@ -0,0 +1,323 @@
+defmodule Algae.Result do
+  @moduledoc ~S"""
+  Data type representing a result.
+  A result can either be a successful `Ok` or an unsuccessful `Err`.
+
+  ## Examples
+
+      iex> load_planet = fn(planet, universe) ->
+      ...>   if planet == "☄️",
+      ...>     do: %Algae.Result.Err{err: "That's not a planet!"},
+      ...>   else: %Algae.Result.Ok{ok: [planet | universe]}
+      ...> end
+      ...>
+      ...> load_planet.("🌏", ["🌕"])
+      %Algae.Result.Ok{ok: ["🌏", "🌕"]}
+      ...> load_planet.("☄️", [])
+      %Algae.Result.Err{err: "That's not a planet!"}
+  """
+
+  import Algae
+  alias Witchcraft.{Chain, Foldable}
+
+  defsum do
+    defdata Err :: any()
```

Comment:
Nice! A direct translation from the Elm version that we discussed, which is great! Have you considered making `Err` a `Exception` struct? That would make it much more compatible with the rest of the ecosystem, Exceptional, Phoenix, &c.

I can see some pros and cons to that, but thought I'd bring it up to start the conversation 😄

Reply:
@expede What were your pros and cons on making this an `Exception.t`?

I'm not really seeing how it could work together, with for example, your 'exceptional' library. Isn't the combination of this Result "monad" and your 'witchcraft' library a replacement for 'exceptional'?

For compatibility with the rest of the ecosystem I currently have the following in my project:
```elixir
def tagged_status_to_result({:ok, a}), do: Result.Ok.new(a)
def tagged_status_to_result({:error, b}), do: Result.Err.new(b)
def tagged_status_to_result(:error), do: Result.Err.new()

def result_to_tagged_status(%Result.Err{err: value}), do: {:error, value}
def result_to_tagged_status(%Result.Ok{ok: value}), do: {:ok, value}
```
(not included in this PR)